### PR TITLE
feat: [CDS-37850]: Endpoint for fetching azure tags

### PR DESCRIPTION
**Pull request by filip5rovic**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)

### CODE REVIEW

**Comment by filip5rovic** [comment]
trigger messagecheck

**Review by vladimir-peric — APPROVED** [review]
> Reviewed 21 of 21 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: :shipit: complete! ✓

**Comment by filip5rovic** [comment]
trigger feature-build

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: Some of the checks are too old. 1 change requests. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/dtos/AzureTagDTO.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3JO35B-IgHOd03b1Px:-N3JO35B-IgHOd03b1Py:b-icent0) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/dtos/AzureTagDTO.java#L28)):*
> ```Java
> @OwnedBy(CDP)
> @Schema(name = "AzureTagResponse", description = "Azure response for tags")
> public class AzureTagDTO {
> ```

This DTO shouldn't be a part of k8s package. 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceService.java` line 19 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3JNhQe-lfmUjECpQvn:-N3JNhQe-lfmUjECpQvo:ba0m9t6) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceService.java#L19)):*
> ```Java
> 
> @OwnedBy(HarnessTeam.CDP)
> public interface AzureResourceService {
> ```

This interface shouldn't be a part of k8s
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceServiceImpl.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3JN5TAAT_WiNwieVhA:-N3JN5TAAT_WiNwieVhB:b-t95qec) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceServiceImpl.java#L43)):*
> ```Java
> @Singleton
> @OwnedBy(HarnessTeam.CDP)
> public class AzureResourceServiceImpl implements AzureResourceService {
> ```

This class should't be a part of k8s. It should be a part of io.harness.cdng.azure

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: Some of the checks are too old. 1 change requests. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/dtos/AzureTagsDTO.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3JOFDc0lxEJTKxKKEd:-N3JOFDd9ZkBUOsGQUxC:b-3jswte) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/dtos/AzureTagsDTO.java#L28)):*
> ```Java
> @OwnedBy(CDP)
> @Schema(name = "AzureTagsResponse", description = "Azure response for tags")
> public class AzureTagsDTO {
> ```

This DTO should't be a part of k8s.

**Comment by filip5rovic** [comment]
trigger feature-build publish-delegate

**Comment by filip5rovic** [comment]
trigger feature-build

**Review by ivan-83 — DISMISSED** [review]
:lgtm:

> Reviewed 3 of 3 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592#-:-N3TWnFR4X_FxPbM6S1L:bnfp4nl)* status: :shipit: complete! ✓

**Review by ivan-83 — DISMISSED** [review]
> Reviewed 8 of 8 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: :shipit: complete! ✓

**Review by ivan-83 — APPROVED** [review]
> Reviewed 3 of 28 files at r8, 25 of 25 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: :shipit: complete! ✓

___
*[`400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java` line 193 at r9](https://reviewable.io/reviews/harness/harness-core/33592#-N3nmekV8BAMuI7xJxEM:-N3nmekV8BAMuI7xJxEN:b-ghu6oi) ([raw file](https://github.com/harness/harness-core/blob/78d2db538d017f8e73cd155895012e22f5bb44bd/400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java#L193)):*
> ```Java
>     io.harness.azure.model.tag.TagDetails tagDetails = new TagDetails();
>     tagDetails.setTagName("tagName");
>     io.harness.azure.model.tag.TagValue tagValue = new TagValue();
> ```

Minor -  remove full qualified names 



_Code quote:_
```Java
io.harness.azure.model.tag
```
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/azureconnector/AzureTaskType.java` line 24 at r9](https://reviewable.io/reviews/harness/harness-core/33592#-N3nu8W-1t1oblKT9iG2:-N3nu8W0DOvHFotnn0EX:bfjaopd) ([raw file](https://github.com/harness/harness-core/blob/78d2db538d017f8e73cd155895012e22f5bb44bd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/connector/azureconnector/AzureTaskType.java#L24)):*
> ```Java
>   LIST_WEBAPP_NAMES,
>   LIST_DEPLOYMENT_SLOTS,
>   LIST_TAGS
> ```

Minor - LIST_TAGS >> LIST_SUBSCRIPTION_TAGS
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/azure/resources/dtos/AzureTagsDTO.java` line 28 at r9](https://reviewable.io/reviews/harness/harness-core/33592#-N3nuaJa3h5krpKidSld:-N3nuaJa3h5krpKidSle:b-6gelii) ([raw file](https://github.com/harness/harness-core/blob/78d2db538d017f8e73cd155895012e22f5bb44bd/125-cd-nextgen/src/main/java/io/harness/cdng/azure/resources/dtos/AzureTagsDTO.java#L28)):*
> ```Java
> @OwnedBy(CDP)
> @Schema(name = "AzureTagsResponse", description = "Azure response for tags")
> public class AzureTagsDTO {
> ```

Minor - just question do you know by the chance why we are using AzureTagsDTO instead of List<AzureTagDTO> i see it is done for all other Azure APIs?

**Comment by filip5rovic** [comment]
trigger ft2

**Comment by filip5rovic** [comment]
trigger ti0

**Review by vladimir-peric — APPROVED** [review]
> Reviewed 23 of 25 files at r9, 3 of 3 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: :shipit: complete! ✓

**Review by ivan-83 — DISMISSED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592#-:-N46NoPtAye5FW-fPV_k:bnfp4nl)* status: :shipit: complete! ✓

File: 400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java [hunk 1]
```diff
@@ -179,13 +182,15 @@ public void testListTags() throws Exception {
 
     AzureManagementRestClient azureManagementRestClient = mock(AzureManagementRestClient.class);
     doReturn(azureManagementRestClient).when(spyAzureDelegateHelperService).getAzureManagementRestClient(any());
-    Call<AzureListTagsResponse> responseCall = (Call<AzureListTagsResponse>) mock(Call.class);
+    Call<io.harness.azure.model.tag.AzureListTagsResponse> responseCall =
```
Comment (puthrayaharness) [inline 400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java:185]: Is the whole path change required?

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java (hunk 1 above)
Comment (filip5rovic) [reply 400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java]: Fixed. For some reason IntelliJ's automation refactoring introduced fqn there

**Review by puthrayaharness — APPROVED** [review]

**Review by raghuAtWings — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/AzureTask.java` line 119 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bL4M1EW7gnvNt5e0:-N46bL4M1EW7gnvNt5e1:buhc1v1) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/AzureTask.java#L119)):*
> ```Java
>       }
>       case LIST_TAGS:
>         validateAzureResourceExist(azureTaskParams, "Could not retrieve any tags because of invalid parameter(s)",
> ```

what happens when the delegate is older version and doesn't have this code. Please test in PR and update.

**Comment by raghuAtWings** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureDelegateHelperService.java` line 29 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bbdiEpQbtnI0ryYi:-N46bbdiEpQbtnI0ryYj:b-xocs8p) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureDelegateHelperService.java#L29)):*
> ```Java
> import io.harness.azure.AzureEnvironmentType;
> import io.harness.azure.model.tag.AzureListTagsResponse;
> import io.harness.azure.model.tag.TagDetails;
> ```

why these new imports if they are not being used?

**Comment by raghuAtWings** [comment]
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureManagementRestClient.java` line 14 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bfG1EFD0HJ5W0ztT:-N46bfG1EFD0HJ5W0ztU:b-2guby3) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureManagementRestClient.java#L14)):*
> ```Java
> import io.harness.annotations.dev.OwnedBy;
> import io.harness.annotations.dev.TargetModule;
> import io.harness.azure.model.tag.AzureListTagsResponse;
> ```

same here

**Review by filip5rovic — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceService.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3IuM6i-qQrfAzB6PNt:-N3PJLNX1ntQE7itReuZ:b-sy69xy) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/125-cd-nextgen/src/main/java/io/harness/cdng/k8s/resources/azure/service/AzureResourceService.java#L28)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

As a part of IdentifierRef connectorRef we already have org and project identifiers, right? 
</blockquote></details>

Yes, they are included
___
*[`930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/AzureTask.java` line 119 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bL4M1EW7gnvNt5e0:-N46bnRN6GbiB5EQ6EG3:b-n2cxgp) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/delegatetasks/azure/AzureTask.java#L119)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

what happens when the delegate is older version and doesn't have this code. Please test in PR and update. 
</blockquote></details>

The new enum won't be used for a couple of months so that the delegate won't be getting unknown enum as param.
I've tested with a new manager and old delegate for existing functionality and it worked as expected.
___
*[`960-api-services/src/main/java/io/harness/azure/impl/AzureManagementClientImpl.java` line 629 at r2](https://reviewable.io/reviews/harness/harness-core/33592#-N3IyRDy15lvt4Dm1DYw:-N3PJSQnAJOiYh_TCocN:b-oz8o77) ([raw file](https://github.com/harness/harness-core/blob/e3bf2715254a6fb1d461efcabeb508e7b076031a/960-api-services/src/main/java/io/harness/azure/impl/AzureManagementClientImpl.java#L629)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

public >> private
</blockquote></details>

Good catch

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 22 of 25 files at r9, 2 of 3 files at r10, 3 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 2. unresolved discussions.

**Review by filip5rovic — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 2. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureDelegateHelperService.java` line 29 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bbdiEpQbtnI0ryYi:-N46cWWg17RxlqfuDdlT:bs7um2t) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureDelegateHelperService.java#L29)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why these new imports if they are not being used?
</blockquote></details>

They are used. It's because those classes have been moved to another package, previously they were in same package so `import` wasn't needed.
___
*[`930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureManagementRestClient.java` line 14 at r12](https://reviewable.io/reviews/harness/harness-core/33592#-N46bfG1EFD0HJ5W0ztT:-N46ckWK5YM6afyjmrG6:bs7um2t) ([raw file](https://github.com/harness/harness-core/blob/2c34abb14f7ef097add58317505b30af06e1d4c0/930-delegate-tasks/src/main/java/software/wings/helpers/ext/azure/AzureManagementRestClient.java#L14)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

same here
</blockquote></details>

They are used. It's because those classes have been moved to another package, previously they were in same package so `import` wasn't needed.

**Review by ivan-83 — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 1. unresolved discussions.

**Review by filip5rovic — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 1. unresolved discussions. 

___
*[`400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java` line 185 at r11](https://reviewable.io/reviews/harness/harness-core/33592#-N3-Te7V1VIqJUuFf1SN-r11-185:-N46fp5i9pDV-PGWevt3:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/df06b158893b7b237332446801a44538f9e45028/400-rest/src/test/java/software/wings/helpers/ext/azure/AzureHelperServiceTest.java#L185)):*
<details><summary><i>Previously, filip5rovic (Filip Petrovic) wrote…</i></summary><blockquote>

Fixed. For some reason IntelliJ's automation refactoring introduced fqn there
</blockquote></details>

Done

**Review by vladimir-peric — APPROVED** [review]
> Reviewed 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592)* status: 1. unresolved discussions.

**Review by raghuAtWings — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33592#-:-N46g7IRDVsDt80XN2sj:bnfp4nl)* status: 1. unresolved discussions.

**Comment by diptiman-adak** [comment]
Merged as per slack discussion

**Pull request merged**